src/values/AbstractObjectValue.js

```diff
@@ -498,21 +501,39 @@ export default class AbstractObjectValue extends AbstractValue {
       let generateAbstractGet = () => {
         let ob = this;
         if (this.kind === "explicit conversion to object") ob = this.args[0];
+        invariant(ob instanceof AbstractValue);
         let type = Value;
         if (P === "length" && Value.isTypeCompatibleWith(this.getType(), ArrayValue)) type = NumberValue;
-        return AbstractValue.createTemporalFromBuildFunction(
+        invariant(typeof P === "string");
+        let obShape = ob.shape;
+        let propertyShape, propertyGetter;
+        if (this.$Realm.instantRender.enabled && obShape !== undefined) {
+          propertyShape = obShape.getMemberAccessShapeInformation(P);
+          if (propertyShape !== undefined) {
+            propertyGetter = obShape.getMemberAccessGraphQLGetter(propertyShape);
+            type = propertyShape.getValueTypeForAbstract();
+          }
+        }
+        let propAbsVal = AbstractValue.createTemporalFromBuildFunction(
           this.$Realm,
           type,
           [ob],
           ([o]) => {
             invariant(typeof P === "string");
-            return memberExpressionHelper(o, P);
+            return propertyGetter !== undefined
+              ? t.callExpression(t.memberExpression(t.identifier("global"), t.identifier("__prop_" + propertyGetter)), [
+                  o,
+                  t.stringLiteral(P),
+                ])
+              : memberExpressionHelper(o, P);
           },
           {
             skipInvariant: true,
             isPure: true,
           }
         );
+        if (propAbsVal instanceof AbstractValue) propAbsVal.shape = propertyShape;
```

Comment:
Instead of this mutation afterwards, you should be able to extend to the `optionsArgs` of `createTemporalFromBuildFunction` to pass this through right away.

Reply:
This means that I need to extend `deriveAbstract`. Also I should handle case for `undefined` in `createTemporalFromBuildFunction`. Are you ok with this? And why there is no handling of null in `createTemporalFromBuildFunction`?

Reply:
If the result type is `undefined` or `null`, there's no point of pushing around shape information, right? Or maybe I didn't understand your question?

Reply:
Hmm, I forgot to actually use my change in this place. My point is that inside `createTemporalFromBuildFunction` has `if (resultType === UndefinedValue) {` but doesn't have this trick for `NullValue`.